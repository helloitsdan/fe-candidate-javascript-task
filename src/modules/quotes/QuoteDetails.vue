<template>
  <div>
    <div class="background--grey-green">
      <div v-if="!quote" class="container padded-block">
        <div v-if="loading" class="loading">Loading...</div>

        <div v-if="error" class="error">
          Sorry! We couldn't load this quote's details right now.
        </div>
      </div>

      <section v-if="quote" class="container padded-block">
        <nav>
          <p>
            <router-link :to="{ name: 'CustomerDetails' }">
              Back to quote list
            </router-link>
          </p>
        </nav>

        <h2>
          Quote for {{ quote.insured_entities | summariseInsuredEntities }}
        </h2>

        <policy-information v-bind="quote" />
      </section>
    </div>

    <div class="background--barely">
      <section v-if="quote" class="container padded-block">
        <policy-entities v-bind="quote" />
      </section>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import PolicyInformation from "@/components/PolicyInformation.vue";
import PolicyEntities from "@/components/PolicyEntities.vue";

import getQuotes from "@/services/quotes";

export default Vue.extend({
  name: "QuoteDetails",
  props: ["quoteId"],
  components: {
    PolicyInformation,
    PolicyEntities
  },
  data() {
    return {
      loading: false,
      error: null,
      quote: null
    };
  },
  created() {
    this.fetchQuoteDetails();
  },
  methods: {
    async fetchQuoteDetails() {
      const { quoteId } = this.$props;

      if (this.loading) {
        return;
      }

      this.loading = true;

      try {
        this.quote = await getQuotes(quoteId);
      } catch (e) {
        this.error = e;
      }

      this.loading = false;
    }
  }
});
</script>
