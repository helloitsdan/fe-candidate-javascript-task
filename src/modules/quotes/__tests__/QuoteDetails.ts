import Vue from "vue";
import { shallowMount } from "@vue/test-utils";

import QuoteDetails from "../QuoteDetails.vue";
import quotesRequest from "@/services/quotes";

jest.mock("@/services/quotes");

describe("QuoteDetails", () => {
  const quotesRequestMock = quotesRequest as jest.Mock;

  const fakeQuote = {
    uuid: "1234",
    status: "ACTIVE",
    start_date: "2016-07-13T02:57:55.235Z",
    insured_entities: [
      { species: "CAT" },
      { species: "CAT" },
      { species: "DOG" }
    ],
  };

  it("renders", async () => {
    quotesRequestMock.mockReturnValue(Promise.resolve(fakeQuote));

    const wrapper = shallowMount(QuoteDetails, {
      stubs: ["router-link"]
    });

    await Vue.nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders a loading message when the request is still pending", () => {
    const wrapper = shallowMount(QuoteDetails, {
      stubs: ["router-link"]
    });
    expect(wrapper.find(".loading").exists()).toBe(true);
  });

  it("renders an error message when the request fails", async () => {
    quotesRequestMock.mockReturnValue(Promise.reject(":("));
    const wrapper = shallowMount(QuoteDetails, {
      stubs: ["router-link"]
    });

    await Vue.nextTick();

    expect(wrapper.find(".error").exists()).toBe(true);
  });
});
