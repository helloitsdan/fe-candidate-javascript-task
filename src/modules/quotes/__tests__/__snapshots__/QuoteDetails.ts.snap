// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteDetails renders 1`] = `
<div>
  <div
    class="background--grey-green"
  >
    <!---->
     
    <section
      class="container padded-block"
    >
      <nav>
        <p>
          <router-link-stub
            to="[object Object]"
          >
            
            Back to quote list
          
          </router-link-stub>
        </p>
      </nav>
       
      <h2>
        
        Quote for 2 cats, 1 dog
      
      </h2>
       
      <policy-information-stub
        insured_entities="[object Object],[object Object],[object Object]"
        start_date="2016-07-13T02:57:55.235Z"
        status="ACTIVE"
        uuid="1234"
      />
    </section>
  </div>
   
  <div
    class="background--clay-dark"
  >
    <section
      class="container padded-block"
    >
      <product-quotes-stub
        insured_entities="[object Object],[object Object],[object Object]"
        start_date="2016-07-13T02:57:55.235Z"
        status="ACTIVE"
        uuid="1234"
      />
    </section>
  </div>
   
  <div
    class="background--barely"
  >
    <section
      class="container padded-block"
    >
      <policy-entities-stub
        insured_entities="[object Object],[object Object],[object Object]"
        start_date="2016-07-13T02:57:55.235Z"
        status="ACTIVE"
        uuid="1234"
      />
    </section>
  </div>
</div>
`;
